resource/interface_factories.cpp.em

```diff
@@ -332,3 +337,197 @@ void ServiceFactory<
+{
+  const uint32_t data_len = vec.size() * sizeof(typename VEC_PRIMITIVE_T::value_type);
+  // copy data from std::vector/std::array into stream
+  memcpy(stream.advance(data_len), &vec.front(), data_len);
+}
+
+// This version is for length
+template<typename VEC_PRIMITIVE_T>
+static void streamPrimitiveVector(ros::serialization::LStream & stream, const VEC_PRIMITIVE_T& vec)
+{
+  const uint32_t data_len = vec.size() * sizeof(typename VEC_PRIMITIVE_T::value_type);
+  stream.advance(data_len);
+}
+
+// This version is for read
+template<typename VEC_PRIMITIVE_T>
+static void streamPrimitiveVector(ros::serialization::IStream & stream, VEC_PRIMITIVE_T& vec)
+{
+  const uint32_t data_len = vec.size() * sizeof(typename VEC_PRIMITIVE_T::value_type);
+  // copy data from stream into std::vector/std::array
+  memcpy(&vec.front(), stream.advance(data_len), data_len);
+}
+
+@[for m in mapped_msgs]@
+
+@[  if m.ros2_msg.package_name=="std_msgs" and m.ros2_msg.message_name=="Header"]
+// msg_2_to_1_stream for std_msgs::(msg)::Header should be specialized because of lack of seq
+@[  else]
+
+template<>
+template<typename STREAM_T, typename ROS2_MSG_T>
+void
+Factory<
+  @(m.ros1_msg.package_name)::@(m.ros1_msg.message_name),
+  @(m.ros2_msg.package_name)::msg::@(m.ros2_msg.message_name)
+>::msg_2_to_1_stream(STREAM_T& stream,
+                     ROS2_MSG_T& ros2_msg)
+{
+@[    if not m.fields_2_to_1]@
+  // No m.fields_1_to_2
+  // TODO : throw exception?
+@[    end if]@
+@[    for ros2_fields, ros1_fields in m.fields_2_to_1.items()]@
+@{
+ros2_field_selection = '.'.join((str(field.name) for field in ros2_fields))
+ros1_field_selection = '.'.join((str(field.name) for field in ros1_fields))
+
+if isinstance(ros2_fields[-1].type, NamespacedType):
+    namespaces = ros2_fields[-1].type.namespaces
+    assert len(namespaces) == 2 and namespaces[1] == 'msg', \
+      "messages not using the '<pkg_name>, msg, <type_name>' triplet are not supported"
+}
+@[      if not isinstance(ros2_fields[-1].type, AbstractNestedType)]@
+  // write non-array field
+@[        if not isinstance(ros2_fields[-1].type, NamespacedType)]@
+  // write primitive field
+  stream.next(ros2_msg.@(ros2_field_selection));
+@[        elif ros2_fields[-1].type.namespaces[0] == 'builtin_interfaces']@
+  // write builtin field
+  ros1_bridge::msg_2_to_1_stream(stream, ros2_msg.@(ros2_field_selection));
+@[        else]@
+  // write sub message field
+  Factory<
+    @(ros1_fields[-1].pkg_name)::@(ros1_fields[-1].msg_name),
+    @(ros2_fields[-1].type.namespaces[0])::msg::@(ros2_fields[-1].type.name)
+  >::msg_2_to_1_stream(stream, ros2_msg.@(ros2_field_selection));
+@[        end if]@
+@[      else]@
+  // write array or sequence field
+@[        if isinstance(ros2_fields[-1].type, AbstractSequence)]@
+  // dynamically sized sequence
+  streamVectorSize(stream, ros2_msg.@(ros2_field_selection));
+@[        else]@
+  // statically sized array
+  // TODO validate ROS1 and ROS2 field sizes match?
```

Comment:
This needs to be fixed before this PR can be accepted.

Reply:
Made fix for this in 2nd commit.